src/comments.ts

```diff
@@ -0,0 +1,297 @@
+        text: string,
+        token: vscode.CancellationToken
+    ): Promise<vscode.CommentThread> {
+        return await replyToCommentThread(document, range, commentThread, text)
+    }
+
+    private didChangeCommentThreads = new vscode.EventEmitter<vscode.CommentThreadChangedEvent>()
+
+    public onDidChangeCommentThreads = this.didChangeCommentThreads.event
+}
+
+async function provideDocumentComments(document: vscode.TextDocument): Promise<vscode.CommentThread[]> {
+    log.appendLine(`provideDocumentComments ${document.uri}`)
+    const [remoteUrl, branch, path] = await repoInfo(document.fileName)
+    if (remoteUrl === '') {
+        throw new Error('Git repository has no remote url configured')
+    }
+    const data = await queryGraphQL(
+        gql`
+            query DiscussionThreads(
+                $targetRepositoryGitCloneURL: String!
+                $targetRepositoryPath: String!
+                $relativeRev: String!
+            ) {
+                discussionThreads(
+                    first: 10000
+                    targetRepositoryGitCloneURL: $targetRepositoryGitCloneURL
+                    targetRepositoryPath: $targetRepositoryPath
+                ) {
+                    totalCount
+                    pageInfo {
+                        hasNextPage
+                    }
+                    nodes {
+                        ...DiscussionThreadFields
+                    }
+                }
+            }
+            ${discussionThreadFieldsFragment}
+        `,
+        {
+            targetRepositoryGitCloneURL: remoteUrl,
+            targetRepositoryPath: path,
+            relativeRev: branch,
+        }
+    )
+
+    if (!data.discussionThreads || !data.discussionThreads.nodes) {
+        throw new Error(`Invalid GraphQL response for DiscussionThreads`)
+    }
+
+    const threads: vscode.CommentThread[] = []
+    for (const thread of data.discussionThreads.nodes) {
+        // TODO: this assumes there is no diff between the document state and the revision
+        const sel = thread.target.relativeSelection
+        if (sel) {
+            const range = new vscode.Range(sel.startLine, sel.startCharacter, sel.endLine, sel.endCharacter)
+            threads.push(discussionToCommentThread(document, range, thread))
+        }
+    }
+
+    return threads
+}
+
+async function createNewCommentThread(
+    document: vscode.TextDocument,
+    range: vscode.Range,
+    text: string
+): Promise<vscode.CommentThread> {
+    log.appendLine(`createNewCommentThread ${document.uri} ${range}`)
+    const [remoteUrl, branch, path] = await repoInfo(document.fileName)
+    if (remoteUrl === '') {
+        throw new Error('Git repository has no remote url configured')
+    }
+
+    const selection = getSelection(document, range)
+
+    const targetRepo: SourcegraphGQL.IDiscussionThreadTargetRepoInput = {
+        repositoryGitCloneURL: remoteUrl,
+        path,
+        branch,
```

Comment:
You need to specify a revision here otherwise as soon as someone pushes a new commit to this repository your threads will become detached from the Git history and it will suck for the web UX later on.